### PR TITLE
feat(app): Persist known robots to file-system when using new discovery

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR adds robot persistence to file to the app and wires up the necessary state. It also includes some tweaks to various values and configurations based on how discovery behaves in the real world.

## changelog

- feat(app): Persist known robots to file-system when using new discovery

## review requests

Try this out on your machine!

0. `make -C discovery-client` (Re-build discovery client)
1. `make -C app dev OT_APP_DISCOVERY__ENABLED=1`

- [x] `discovery.robotsByName` state is prepopulated at app launch from file
- [x] App picks up virtual smoothie as `opentrons-dev`
- [x] App picks up IPv6 robots as their actual name
- [x] Unplugging an IPv6 robot from USB will get it to fall back to WiFi (faster if list is currently refreshing)
- [x] Ditto for IPv4 (should be unchanged from #2045)
- [x] RPC connection to robots still works

Known issues:

- RPC client and health-check module **do not** handle USB disconnections gracefully when RPC is connected
    - This may be a blocker for actual release of new discovery
    - For now though its enough for me to keep it behind the feature flag

### CODE REVIEW

File: app/scripts/advertise-local-api.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // simple script to advertise a MDNS service for local API
 // TODO(mc, 2017-10-31): remove this file once API can advertise for itself
```
Comment (mcous) [inline app/scripts/advertise-local-api.js]: Technically should've updated this TODO. This file will now go away when new discovery is enabled by default

File: app/src/discovery/index.js [hunk 2]
```diff
@@ -99,32 +111,40 @@ export function discoveryReducer (
...
-          [robot.name]: robot
-        }), {})
+        robotsByName: flattenRobots(action.payload.robots)
       }
   }
 
   return state
 }
 
+function flattenRobots (robots: Array<DiscoveredRobot> = []): RobotsMap {
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/discovery/index.js]: You're right, I'm gonna rename this to `normalizeRobots` as per redux convention. Don't know what was happening in my head when I named it "flatten"

File: app/src/discovery/index.js [hunk 3]
```diff
@@ -99,32 +111,40 @@ export function discoveryReducer (
...
+  return robots.reduce(
+    (robotsMap, robot) => ({
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/discovery/index.js]: Noted. Already touching this file; will add type annotations for safety

**Comment by Laura-Danielle** [comment]
A few comments from Windows 10:

Took a few refreshes to see the robot USB (IPv4) -> could be what that customer was seeing earlier today.
Same with getting the robot to show up for WiFi. IPv6 worked like a 🥇

**Pull request merged**